### ISSUE TITLE
Torch-MLIR's RNGs result in highly correlated random tensors

### ISSUE THREAD

**Issue by ramiro050** [llvm/torch-mlir#1608]
The current implementation of RNGs in Torch-MLIR creates random tensors that are highly correlated to one another. For example, running the script below that produces two uniform tensors results in the following correlation coefficient matrix for the two tensors:

```
tensor([[1.0000, 0.9701],
        [0.9701, 1.0000]], dtype=torch.float64)
```

In contrast, PyTorch produces the following matrix:

```
tensor([[ 1.0000e+00, -4.2704e-05],
        [-4.2704e-05,  1.0000e+00]], dtype=torch.float64)
```

My suspicion is that because the `getNextSeed` op and the implementation of `ConvertAtenUniformOp` use the same multiplier and increment to generate the next random number, this results in a high degree of correlation from one uniform tensor to the next.

This correlation prevents [certain decompositions](https://github.com/llvm/torch-mlir/pull/1581#discussion_r1023602897) in Torch-MLIR from working correctly.

## Script to recreate issue

```python
import torch
import torch_mlir
from torch_mlir_e2e_test.linalg_on_tensors_backends.refbackend \
    import RefBackendLinalgOnTensorsBackend

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y):
        u1 = x.uniform_()
        u2 = y.uniform_()
        return u1, u2

example_inputs = [torch.empty((1000000), dtype=torch.float64),
                  torch.empty((1000000), dtype=torch.float64)]
module = MyModule()
script_module = torch.jit.script(module, example_inputs)
torch_mlir_module = torch_mlir.compile(script_module, example_inputs,
                                       output_type="linalg-on-tensors")

backend = RefBackendLinalgOnTensorsBackend()
compiled = backend.compile(torch_mlir_module)
jit_module = backend.load(compiled)

print("Running TorchScript Graph")
print('Correlation coefficient matrix:')
script_results = script_module.forward(*example_inputs)
print(torch.corrcoef(torch.stack(script_results, 0)))
print("Running Compiled Graph")
print('Correlation coefficient matrix:')
numpy_inputs = list(map(lambda x: x.numpy(), example_inputs))
jit_results = tuple(map(lambda x: torch.tensor(x), jit_module.forward(*numpy_inputs)))
print(torch.corrcoef(torch.stack(jit_results, 0)))
```

**Comment by ramiro050** [comment]
I have implemented in Python the RNG algorithm used in torch-mlir to make it easier to see what's happening and experiment (this led to catching the bug fixed in https://github.com/llvm/torch-mlir/pull/1625, which improves correlation from 0.97 to 0.5). 

I've tried playing around with different `MULTIPLIER` values, but that has not led to any significant improvements. 

The main limitation with the current implementation seems to be that the code that generates a uniformly random tensor uses the same seed, `initial_seed` in `uniform`, to generate each of the values in the tensor. The algorithm does take into account the indices where each value will be stored in order to have different values everywhere in the tensor. However, this means that when two uniform tensors of the same size are generated, since the only thing that changes is the `initial_seed` from one constant to another, the values at a given index in both tensors follow a similar path when being generated, leading to high correlation. This is especially problematic when generating rank-1 tensors, where each new tensor is equal to the previous tensor plus a constant (modulo max value).

In order to avoid the current limitation mentioned, we would need to use a different seed for every call to `get_next_temp`. Doing this in the script below decreases the correlation between the two tensors generated to `-0.0001`. Moreover, this is the reason why the implementation in https://github.com/llvm/torch-mlir/pull/1581 works well. Because the two uniform variables needed to generate the normal variable are generated *sequentially* in the `linalg::GenericOp`,

 https://github.com/llvm/torch-mlir/blob/690e9a3ad81f670c17f83b6fec31bc7acf204bf7/lib/Conversion/TorchToLinalg/Random.cpp#L249-L255 

the two uniform variables are always starting with very different initial seeds, and the initial seed used for the second uniform variable varies from one index to the next. This type of behavior where the two uniform variables used for the normal variable have to different initial seeds and one of the uniform variables has an initial seed that varies across the entire tensor cannot be achieved as a decomposition with the current algorithm for RNGs. 

The main problem at the moment is that I don't know how we can modify the algorithm to use a different seed when calling `get_next_temp` while preserving it's parallelism properties. 

### LCG Algorithm script

```python
from typing import Iterable, Tuple
import numpy as np

MULTIPLIER = 6364136223846793005
INCREMENT = 1442695040888963407

_global_seed = 0
def get_next_seed() -> int:
    global _global_seed
    _global_seed = (_global_seed * MULTIPLIER + INCREMENT) % (2**64)
    return _global_seed

def get_next_temp(seed: int, indices: Iterable[int]) -> int:
    curr_temp = seed
    for index in indices:
        curr_temp = ((index + curr_temp) * MULTIPLIER + INCREMENT) % (2**64)
    return curr_temp

def uniform(size: Tuple[int]) -> np.array:
    result = np.empty(size)
    rand_max = 2**64 - 1
    initial_seed = get_next_seed()
    for indices in np.ndindex(size):
        random_val = get_next_temp(initial_seed, indices)
        result[indices] = random_val / rand_max
    return result

shape = (10, 10, 10, 10, 10)
u1 = uniform(shape).flatten()
u2 = uniform(shape).flatten()
stack = np.stack([u1, u2])
print(np.corrcoef(stack))
```

**Comment by silvasean** [comment]
I started a discussion here: https://discourse.llvm.org/t/centralized-place-for-smart-fast-correct-enough-parallel-computation-stuff/66744
